packages/workspace/src/workspace/nacl_files/multi_env/routers.ts

```diff
@@ -212,10 +212,22 @@ export const routeDefault = async (
   // However - if we have only one env and the user moved the top level element to be
   // env specific - we respect that and add the change to the env.
   if (change.action === 'add') {
-    const primTopLevelElement = await primarySource.get(change.id.createTopLevelParentID().parent)
-    return _.isEmpty(secondarySources) && primTopLevelElement === undefined
-      ? { commonSource: [change] }
-      : { primarySource: [change] }
+    const topLevelID = change.id.createTopLevelParentID().parent
+    const primTopLevelElement = await primarySource.get(topLevelID)
+    const commonTopLevelElement = await commonSource.get(topLevelID)
+    const secTopLevelElements = await Promise.all(
+      Object.values(secondarySources).map(src => src.get(topLevelID))
+    )
+    const hasPrimTopLevel = primTopLevelElement !== undefined
+    const hasCommonTopLevel = commonTopLevelElement !== undefined
+    const hasSecTopLevel = _.some(secTopLevelElements, srcElem => srcElem !== undefined)
+    if (_.isEmpty(secondarySources) && !hasPrimTopLevel) {
+      return { commonSource: [change] }
+    }
+    if (hasCommonTopLevel && !hasPrimTopLevel && !hasSecTopLevel) {
+      return { commonSource: [change] }
+    }
+    return { primarySource: [change] }
```

Comment:
readability nit: I think it might be worth naming some of these based on the "conclusion" instead of the underlying condition - it took me some time to understand that the reason we need two conditions for `common` is that if the element is completely new and there are no secondary sources we still want it to go to common. maybe something like
```suggestion
    const shouldPreferCommon = async (elemID: ElemID): Promise<boolean> => (
      (_.isEmpty(secondarySources) || (await commonSource.get(elemID) !== undefined))
    )
    const elementExistsInEnvs = async (elemID: ElemID): Promise<boolean> => {
      const topLevelElements = await Promise.all(
        Object.values(secondarySources).concat(primarySource).map(src => src.get(elemID))
      )
      return topLevelElements.some(srcElem => srcElem !== undefined)
    }

    const topLevelID = change.id.createTopLevelParentID().parent
    if (await shouldPreferCommon(topLevelID) && !(await elementExistsInEnvs(topLevelID))) {
      return { commonSource: [change] }
    }
    return { primarySource: [change] }
```
? (hope I didn't mess up any of the conditions 🤞 )

Reply:
Changes a bit - lemme know if you think its better...

Reply:
:+1: thanks